test/integration/integration_test.go

```diff
@@ -505,6 +506,13 @@ func (i *integrationTest) exerciseIncrementalBuild(tag, imageName string, remove
 	if checkOnBuild {
 		i.fileExists(containerID, "/sti-fake/src/onbuild")
 	}
+
+	if took := time.Since(start); took > docker.DefaultDockerTimeout {
+		// https://github.com/openshift/source-to-image/issues/301 is a
+		// case where incremental builds would get stuck until the
+		// timeout.
+		t.Errorf("Test took too long (%v), some operation may have gotten stuck waiting for the DefaultDockerTimeout (%v). Inspect the logs to find operations that took long.", took, docker.DefaultDockerTimeout)
+	}
```

Comment:
Out of scope. Somewhere deep inside I feel `DefaultDockerTimeout` started as just a hack to accommodate synchronization issues. But it can come in handy when operating against a faulty docker daemon.

Worry: such an integration test would be meaningful, but notably slow. Another possibility would be unit tests around the blocks that use the docker client; temporarily decrease `DefaultDockerTimeout`, pass in a fake slow client, then ensure the calls time out.

> at every step carries out actions which run for longer than DefaultDockerTimeout, which should always succeed

You mean should always fail? If they run longer than the timeout, they should be aborted and fail.